src/v/cloud_storage/cache_service.cc

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_storage/logger.h"
+#include "random/generators.h"
+#include "utils/gate_guard.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/file.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/seastar.hh>
+
+#include <cloud_storage/cache_service.h>
+
+#include <exception>
+#include <optional>
+
+namespace cloud_storage {
+
+cache::cache(
+  std::filesystem::path cache_dir,
+  size_t max_cache_size,
+  ss::lowres_clock::duration check_period) noexcept
+  : _cache_dir(std::move(cache_dir))
+  , _max_cache_size(max_cache_size)
+  , _check_period(check_period) {}
+
+ss::future<> cache::start() {
+    // todo: start maintenance cycle here
+    vlog(cst_log.debug, "max_cache_size: {}", _max_cache_size);
+    vlog(cst_log.debug, "check_period: {}", _check_period.count());
+    return ss::make_ready_future<>();
+}
+
+ss::future<> cache::stop() { return _gate.close(); }
+
+ss::future<std::optional<cache_item>> cache::get(std::filesystem::path key) {
+    gate_guard guard{_gate};
+    ss::file cache_file;
+    try {
+        cache_file = co_await ss::open_file_dma(
+          (_cache_dir / key).native(), ss::open_flags::ro);
+    } catch (...) {
+        vlog(
+          cst_log.debug,
+          "Couldn't open file {} to read: {}",
+          (_cache_dir / key).native(),
+          std::current_exception());
+        co_return std::nullopt;
+    }
+
+    auto data_size = co_await cache_file.size();
+    auto data_stream = ss::make_file_input_stream(cache_file);
+    co_return std::optional(cache_item{std::move(data_stream), data_size});
+}
+
+ss::future<>
+cache::put(std::filesystem::path key, ss::input_stream<char> data) {
+    gate_guard guard{_gate};
+
+    auto filename = (_cache_dir / key).filename();
+    auto tmp_filename = std::filesystem::path(
+      filename.string() + random_generators::gen_alphanum_string(4).data()
+      + ".part");
+    auto dir_path = (_cache_dir / key).remove_filename();
+
+    ss::output_stream<char> out;
+    try {
+        co_await ss::recursive_touch_directory(dir_path.string());
+
+        auto flags = ss::open_flags::rw | ss::open_flags::create
+                     | ss::open_flags::truncate;
```

Comment:
i don't know how the cache is going to be used, but is there ever any risk that a file is truncated that is also opened for reading at the same time (e.g. via `get`)? is there a way to check for this condition and abort if it occurs?

Reply:
I think that this shouldn't be a problem. If I understand correctly `O_CREATE` will allocate new inode for the file and the file handle opened in `get` will be using an old inode.

And because of the way it's going to be used I don't think this will ever happen. We can potentially have a concurrent `put` calls. But we shouldn't have concurrent `put` and `get` on the same file.

Reply:
@Lazin 

`I think that this shouldn't be a problem. If I understand correctly O_CREATE will allocate new inode for the file and the file handle opened in get will be using an old inode.`

this is only true when the file is deleted: existing open file handles will continue to work with the previous instance of the file. if you truncate an open file it will be observed by other open handles to the same file.

`We can potentially have a concurrent put calls` if it is true that concurrent puts to the same file are allowed then it looks to me like you will have corruption.

especially in a cache i think it is worth being stricter than this. a semaphore or assertions could be used to guarantee that concurrent get and put do not occur. debugging a cache issue later may be really hard. so perhaps a mutex to protect mutations.

Reply:
I thought that I replied via email but I don't see my reply here. Anyway, you're probably right about O_CREATE but we're creating files with randomized names in put. So it should work OK without extra synchronization. There is slight chance that we will have concurrent calls to put and the names will clash but I think that this can be mitigated by using simple counter instead of the random value in the temporary file path.

Reply:
Instead of using `random_generators::gen_alphanum_string(4)` it should be possible to just have an atomic counter somewhere which is incremented on every `put` call. And in the path of the temporary we can use this number instead of random thing.